File: boot/serial.c
/*
 * serial.c — Minimal 16550-compatible UART driver implementation.
 * See serial.h and ADR-004 for rationale.
 *
 * Register offsets from SERIAL_COM1_BASE (standard 16550 layout):
 *   +0  DLAB=0: Data register (read/write)      DLAB=1: Divisor LSB
 *   +1  DLAB=0: Interrupt Enable Register        DLAB=1: Divisor MSB
 *   +2  FIFO Control Register (write) / Interrupt Identification (read)
 *   +3  Line Control Register (sets DLAB in bit 7)
 *   +4  Modem Control Register
 *   +5  Line Status Register
 */

#include "include/serial.h"

#define REG_DATA          0
#define REG_INT_ENABLE    1
#define REG_FIFO_CTRL     2
#define REG_LINE_CTRL     3
#define REG_MODEM_CTRL    4
#define REG_LINE_STATUS   5

#define LCR_8N1        0x03 /* 8 data bits, no parity, 1 stop bit */
#define LCR_DLAB       0x80 /* Divisor Latch Access Bit */

#define FCR_ENABLE_FIFO       0x01
#define FCR_CLEAR_RX_FIFO     0x02
#define FCR_CLEAR_TX_FIFO     0x04
#define FCR_14BYTE_TRIGGER    0xC0

#define MCR_DTR   0x01
#define MCR_RTS   0x02
#define MCR_OUT2  0x08 /* must be set for interrupts on real hardware;
                           harmless and conventional to set even though
                           this driver runs polled, not interrupt-driven */

#define LSR_TX_EMPTY 0x20 /* Transmit Holding Register Empty */

/* UART input clock is 1.8432 MHz on standard PC hardware; the baud
 * rate divisor is that clock divided by (desired baud * 16). For our
 * target of 38400 baud: 1843200 / (38400 * 16) = 3. This is the
 * well-known standard divisor value for 38400 baud on 16550 hardware,
 * derived from the UART's documented clock, not copied from any OS
 * driver source. */
#define BAUD_DIVISOR_38400 3

/*
 * outb/inb — single-byte port I/O via inline assembly.
 *
 * "dN" constrains the port number to the DX register (required by the
 * x86 IN/OUT instruction encoding for 16-bit port numbers above 0xFF,
 * which SERIAL_COM1_BASE's port and its offsets are not, but using dN
 * uniformly keeps this correct if a port ever needs relocating above
 * 0xFF). "a" constrains the data value to AL, per the instruction's
 * fixed accumulator-register encoding.
 */
static inline void outb(UINT16 port, UINT8 value)
{
    __asm__ __volatile__("outb %0, %1" : : "a"(value), "Nd"(port));
}

static inline UINT8 inb(UINT16 port)
{
    UINT8 value;
    __asm__ __volatile__("inb %1, %0" : "=a"(value) : "Nd"(port));
    return value;
}

void SerialInit(void)
{
    /* Disable all UART-generated interrupts — this driver is polled,
     * and we have not set up an interrupt controller or handlers at
     * this point in boot (that is Phase 3 kernel work), so any
     * interrupt this UART raised would have nowhere correct to go. */
    outb(SERIAL_COM1_BASE + REG_INT_ENABLE, 0x00);

    /* Enable DLAB to expose the divisor latch registers at offsets
     * 0 and 1, program the divisor for 38400 baud, then clear DLAB
     * again so offset 0 goes back to being the data register. */
    outb(SERIAL_COM1_BASE + REG_LINE_CTRL, LCR_DLAB);
    outb(SERIAL_COM1_BASE + REG_DATA, (UINT8)(BAUD_DIVISOR_38400 & 0xFF));
    outb(SERIAL_COM1_BASE + REG_INT_ENABLE, (UINT8)((BAUD_DIVISOR_38400 >> 8) & 0xFF));
    outb(SERIAL_COM1_BASE + REG_LINE_CTRL, LCR_8N1);

    /* Enable and reset the FIFOs, with a 14-byte receive trigger level
     * (irrelevant for our transmit-only usage, but this is the
     * standard/correct full initialization sequence for real 16550
     * hardware rather than a cut corner). */
    outb(SERIAL_COM1_BASE + REG_FIFO_CTRL,
         FCR_ENABLE_FIFO | FCR_CLEAR_RX_FIFO | FCR_CLEAR_TX_FIFO | FCR_14BYTE_TRIGGER);

    /* Assert DTR/RTS and OUT2 — standard modem control bring-up so a
     * real UART (and QEMU's emulation of one) considers the line
     * "ready," matching how real hardware initialization sequences
     * are documented. */
    outb(SERIAL_COM1_BASE + REG_MODEM_CTRL, MCR_DTR | MCR_RTS | MCR_OUT2);
}

void SerialWriteChar(char c)
{
    if (c == '\n') {
        SerialWriteChar('\r');
    }

    /* Poll the Line Status Register until the Transmit Holding
     * Register is empty (bit 5) — the correct, non-shortcut way to
     * avoid overwriting a byte the UART hasn't shifted out yet.
     * There is no interrupt-driven alternative available yet (no
     * interrupt handling exists this early in boot), so polling is
     * not a corner cut here, it is the only correct option available
     * at this stage of the system's life. */
    while ((inb(SERIAL_COM1_BASE + REG_LINE_STATUS) & LSR_TX_EMPTY) == 0) {
        /* busy-wait */
    }

    outb(SERIAL_COM1_BASE + REG_DATA, (UINT8)c);
}

void SerialWriteString(const CHAR8 *str)
{
    while (*str != '\0') {
        SerialWriteChar(*str);
        str++;
    }
}

void SerialWriteHex64(UINT64 value)
{
    static const char hexDigits[] = "0123456789ABCDEF";
    char buffer[17]; /* 16 hex digits + null terminator */

    buffer[16] = '\0';
    for (int i = 15; i >= 0; i--) {
        buffer[i] = hexDigits[value & 0xF];
        value >>= 4;
    }

    SerialWriteString(buffer);
}

File: kernel/src/arch/x86_64/serial.rs
//! serial.rs — minimal 16550-compatible UART driver, COM1 only, for
//! early kernel boot diagnostics.
//!
//! Reimplemented independently from `boot/serial.c` rather than
//! shared — ADR-004 already established that the bootloader's serial
//! handling is not linked into the kernel; the kernel owns its own
//! device state from the moment it starts (the same principle
//! `tests/kernel_stub/kernel_stub.c` demonstrated by re-initializing
//! the UART even though the bootloader had just done so). This is
//! explicitly NOT the formal driver model Phase 4 will define — see
//! ADR-006's "Early debug output vs. Phase 4 drivers".
//!
//! Register layout and initialization sequence are the same
//! documented 16550 hardware behavior `boot/serial.c` used — an
//! external hardware specification, implemented independently in a
//! second language, not code shared or copied between the two.

const COM1: u16 = 0x3F8;

const REG_DATA: u16 = 0;
const REG_INT_ENABLE: u16 = 1;
const REG_FIFO_CTRL: u16 = 2;
const REG_LINE_CTRL: u16 = 3;
const REG_MODEM_CTRL: u16 = 4;
const REG_LINE_STATUS: u16 = 5;

const LCR_8N1: u8 = 0x03;
const LCR_DLAB: u8 = 0x80;

const FCR_ENABLE_FIFO: u8 = 0x01;
const FCR_CLEAR_RX_FIFO: u8 = 0x02;
const FCR_CLEAR_TX_FIFO: u8 = 0x04;
const FCR_14BYTE_TRIGGER: u8 = 0xC0;

const MCR_DTR: u8 = 0x01;
const MCR_RTS: u8 = 0x02;
const MCR_OUT2: u8 = 0x08;

const LSR_TX_EMPTY: u8 = 0x20;

/// 1.8432 MHz UART clock / (38400 baud * 16) = 3 — the standard 16550
/// divisor for 38400 baud, matching `boot/serial.c`'s choice so a
/// terminal watching the boot process doesn't need to change baud
/// rate mid-stream when control passes from bootloader to kernel.
const BAUD_DIVISOR_38400: u16 = 3;

/// SAFETY: `out`/`in` on a fixed, always-valid I/O port number is sound
/// as a primitive operation — the actual safety burden (not racing
/// with another core, not doing this before paging/interrupts are in
/// a state that allows it) is on the caller, which is why these stay
/// private to this module and only `init`/`write_*` are exposed.
unsafe fn outb(port: u16, value: u8) {
    core::arch::asm!("out dx, al", in("dx") port, in("al") value, options(nomem, nostack, preserves_flags));
}

unsafe fn inb(port: u16) -> u8 {
    let value: u8;
    core::arch::asm!("in al, dx", out("al") value, in("dx") port, options(nomem, nostack, preserves_flags));
    value
}

/// Program the UART for 38400 8N1 with FIFOs enabled. Must be called
/// before any `write_*` function — matches `SerialInit()` in
/// `boot/serial.c` field-for-field; see that file's comments for why
/// each step is necessary rather than assumed.
pub fn init() {
    // SAFETY: COM1's I/O ports are fixed, architecturally-defined
    // addresses; this sequence is the standard documented 16550
    // initialization procedure, called once at kernel boot before any
    // interrupt handling exists to race with it.
    unsafe {
        outb(COM1 + REG_INT_ENABLE, 0x00);

        outb(COM1 + REG_LINE_CTRL, LCR_DLAB);
        outb(COM1 + REG_DATA, (BAUD_DIVISOR_38400 & 0xFF) as u8);
        outb(COM1 + REG_INT_ENABLE, ((BAUD_DIVISOR_38400 >> 8) & 0xFF) as u8);
        outb(COM1 + REG_LINE_CTRL, LCR_8N1);

        outb(
            COM1 + REG_FIFO_CTRL,
            FCR_ENABLE_FIFO | FCR_CLEAR_RX_FIFO | FCR_CLEAR_TX_FIFO | FCR_14BYTE_TRIGGER,
        );

        outb(COM1 + REG_MODEM_CTRL, MCR_DTR | MCR_RTS | MCR_OUT2);
    }
}

fn write_byte(byte: u8) {
    if byte == b'\n' {
        write_byte(b'\r');
    }

    // SAFETY: polling the Line Status Register before writing is the
    // documented-correct way to avoid overwriting a byte the UART
    // hasn't shifted out yet; no interrupt-driven alternative exists
    // this early in kernel boot (no IDT installed yet).
    unsafe {
        while (inb(COM1 + REG_LINE_STATUS) & LSR_TX_EMPTY) == 0 {}
        outb(COM1 + REG_DATA, byte);
    }
}

/// Write a UTF-8 string. Only the ASCII subset is meaningful over a
/// raw serial terminal in this early-boot context, but taking `&str`
/// (not `&[u8]`) keeps call sites using ordinary Rust string literals
/// rather than byte-string literals throughout the kernel.
pub fn write_str(s: &str) {
    for byte in s.bytes() {
        write_byte(byte);
    }
}

/// Write a `u64` as a fixed-width 16-digit hex string — mirrors
/// `SerialWriteHex64` in `boot/serial.c`, for the same reason
/// (addresses and sizes at these magnitudes are meaningless in
/// decimal, and there is no `println!`/formatting machinery this
/// early — `core::fmt` integration is a natural follow-up once the
/// kernel heap exists, not implemented speculatively now).
pub fn write_hex64(value: u64) {
    const HEX_DIGITS: &[u8; 16] = b"0123456789ABCDEF";
    let mut buf = [0u8; 16];
    let mut v = value;
    for slot in buf.iter_mut().rev() {
        *slot = HEX_DIGITS[(v & 0xF) as usize];
        v >>= 4;
    }
    for &b in &buf {
        write_byte(b);
    }
}
